operator/watchers/service_cache.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchers
+
+import (
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+func newServiceCacheSubscriber(swgSvcs, swgEps *lock.StoppableWaitGroup) *serviceCacheSubscriber {
+	return &serviceCacheSubscriber{
+		swgSvcs: swgSvcs,
+		swgEps:  swgEps,
+	}
+}
+
+// serviceCacheSubscriber represents an object that's subscribed to K8s service
+// events in order to keep the K8sSvcCache up-to-date. It implements
+// subscriber.ServiceHandler and is used in the watcher.
+type serviceCacheSubscriber struct {
+	swgSvcs, swgEps *lock.StoppableWaitGroup
+}
+
+func (c *serviceCacheSubscriber) OnAdd(obj *slim_corev1.Service) {
+	log.WithField(logfields.ServiceName, obj.Name).Debugf("Received service addition %+v", obj)
+	K8sSvcCache.UpdateService(obj, c.swgSvcs)
```

Comment:
`K8sSvcCache.UpdateService` sends the events to a buffered channel of some pre-configured capacity. Currently, we have only one buffered channel as `K8sSvcCache.UpdateService` is directly called from the watcher callback. But with this change there'll be a channel for every subscriber. Won't that increase memory usage of Cilium? Also, it looks like the [code](https://github.com/cilium/cilium/blob/master/pkg/k8s/service_cache.go#L183..201) to parse the `Service` object, call `correlateEndpoints`, etc in `K8sSvcCache.UpdateService` be duplicated for every subscriber?

Reply:
No, that's not the intent. I opened https://github.com/cilium/cilium/issues/15471 to further address subscribers that depend on the service cache itself. An example is explained there.

Generally , the solution wouldn't be to create a new channel for each subscriber, but rather refactor the downstream subscribers that _depend_ on the service cache itself.